bluemira/power_cycle/tools.py

```diff
@@ -31,3 +34,93 @@ def create_axes(ax=None):
     if ax is None:
         _, ax = plt.subplots()
     return ax
+
+
+def unique_domain(x, y, epslon=1e-10, mode="careful"):
```

Comment:
Answer to each of your points:
- y is passed to the function to ensure that x and y are a valid pair of domain/image vectors; often you will encounter vectors of different lengths that need to be adjusted, for example by padding one of them. I have just not implemented that yet due to lack of time.
- the slow version is "more correct" just because it does not risk accumulating a nudge that becomes significant when you're dealing with many points... but running it becomes unfeasible when dealing with domain/image sets in the order of hundreds of millions of points that must be down-sampled, especially if you have multiple sets... that is exactly what happened with the original coil supply system data I got... so I still think the fast mode is necessary

Reply:
1) isnt that what your match_domains does or is meant to do?
2) I'm still on the side of disagree but if you _really_ need it its a kw-only boolean flag that warns when used:
```python
if flag:
    bluemira_warn('be careful with this')  # be slightly more descriptive but should fit on a line
else:
    # normal version
```

Reply:
also, the use of pairwise breaks the case in which len(x) = 1; the function should just return the same x

Reply:
then the function doesnt do anything and should return early

Reply:
> the mode needs to be selected more cleanly. We also do not need to provide short hand selections. The below enum will cover all needs and also remove the need for the ValueError raise in the function.
> 
> ```python
> from enum import Enum, auto
> 
> class UniqueDomainMode(Enum):
>     CAREFUL = auto()
>     QUICK = auto()
>     
>     @classmethod
>     def _missing_(cls, value: str):
>         try:
>             return cls[value.upper()]
>         except KeyError:
>             raise AttributeError(...)
> ```

can we add the [Advanced Enum package](https://pypi.org/project/aenum/) to the requirements? I'd like to use multi-valued enums.

Reply:
> 1. isnt that what your match_domains does or is meant to do?

exactly: I was concerned about some edge cases in which one might not call `match_domains`... so I was precisely debating whether to shift this particular result of `interp` into `unique_domain`, and have it called twice... but I have not decided on this yet; seems inefficient to call the same function twice.




> 2. I'm still on the side of disagree, millions of points is not our domain. If you _really_ need it its a kw-only boolean flag that warns when used:
> ```python
> if flag:
>     bluemira_warn('be careful with this')  # be slightly more descriptive but should fit on a line
> else:
>     # normal version
> ```
> 
> I'm still not sold on adding it

millions of points are not calculated by our models, but it is necessary for dealing with data files for testing, verification and generating input files...
when you talk about a flag, you mean adding the fast mode and only use it with a "hidden" key in a kwargs?

Reply:
See https://docs.astral.sh/ruff/rules/boolean-default-value-positional-argument/ for boolean args
```python
def func(a: float, *, b: bool = False):
    print(a, b)

func(1, False) # Will error
func(1, b=False) # prints 1 False
```

Reply:
If its only for testing and verification dependent on the time it takes, that seems like overkill for our purposes.

Reply:
Also if you use numba its not slow:

```python
import numba as nb

@nb.jit
def unique_domain(x, epslon=1e-10):
    """
    Ensure x has only unique values to make (Domain: x -> Image: y) a function.

    Epslon must be small enough so that consecutive values can be considered
    equal within the context/scale in which the function is defined.

    Careful mode:
    -------------
    Nudge forward each non-unique element in x by (N * epslon), given N times,
    after the first appearance, that value has appeared in x before.

    Fast mode:
    ---------
    Nudge forward every element in x by (N * epslon), with N being the index
    of that element in x. More prone to errors if the number of elements in
    x is large.
    """
    new_x = [x[0]]
    nudge = 0
    for x_last, x_this in zip(x[:-1], x[1:]):
        if np.isclose(x_last, x_this, rtol=EPS):
            nudge += epslon
            x_this = x_last + nudge
        else:
            nudge = 0
        new_x.append(x_this)
    return np.asarray(new_x)
```
10M points take 105ms

![image](https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/81617086/0fac1b6c-589f-4214-8b26-80dcd7e2eaf9)

Reply:
> See https://docs.astral.sh/ruff/rules/boolean-default-value-positional-argument/ for boolean args

I still don't get your point, you suggested the boolean flag but why can't we have a keyword argument with a default value? it will never be modified, unless necessary.

Reply:
as per my comment above we dont need fast because slow is fast. Secondly there is only two options fast_on/fast_off and really we dont want people to use the fast option regularly therefore the req of keyword. Thirdly string inputs are a bad design pattern.

Reply:
> 10M points take 105ms, if you're using more points than that you need to preprocess the data properly

"pre-process the data properly" is precisely the problem, it is [extremely non-trivial](https://www.mathworks.com/matlabcentral/answers/1454924-downsample-data-adapively-intelligently#answer_789429) and there are multiple potential algorithms that may or may not work... "proper" pre-processing is a type of optimization and that is an open field

this function is what I developed for when someone sends me data from system X in plant design, sometimes with hundreds of points, sometimes with hundreds of millions (_e.g._ data from control engineering)...

if it takes too long to just prepare a representative set of data points before even starting the down-sampling algorithms for comparison, the fast mode in matching the domain of vectors helps a lot

Reply:
Just to confirm my point of not needing fast. Fast is only 10% faster while also possibly being wrong. Thats not a good enough reason to include it.

![image](https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/81617086/a8d49558-5295-4005-9ea4-05115e283ef4)